homeassistant/components/tesla/sensor.py

```diff
@@ -17,89 +19,101 @@
-        if self.type:
-            return f"{self.tesla_id}_{self.type}"
-        return self.tesla_id
+        return (
+            super().unique_id if not self.type else f"{super().unique_id}_{self.type}"
+        )
 
     @property
-    def state(self):
+    def state(self) -> Optional[float]:
         """Return the state of the sensor."""
-        return self.current_value
+        if self.tesla_device.type == "temperature sensor":
+            if self.type == "outside":
+                return self.tesla_device.get_outside_temp()
+            return self.tesla_device.get_inside_temp()
+        if self.tesla_device.type in ["range sensor", "mileage sensor"]:
+            units = self.tesla_device.measurement
+            if units == "LENGTH_MILES":
+                return self.tesla_device.get_value()
+            return round(
+                convert(self.tesla_device.get_value(), LENGTH_MILES, LENGTH_KILOMETERS),
+                2,
+            )
+        if self.tesla_device.type == "charging rate sensor":
+            return self.tesla_device.charging_rate
+        return self.tesla_device.get_value()
 
     @property
-    def unit_of_measurement(self):
+    def unit_of_measurement(self) -> Optional[str]:
         """Return the unit_of_measurement of the device."""
-        return self.units
+        units = self.tesla_device.measurement
+        if units == "F":
+            return TEMP_FAHRENHEIT
+        if units == "C":
+            return TEMP_CELSIUS
+        if units == "LENGTH_MILES":
+            return LENGTH_MILES
+        if units == "LENGTH_KILOMETERS":
+            return LENGTH_KILOMETERS
+        return units
 
     @property
-    def device_class(self):
+    def device_class(self) -> Optional[str]:
         """Return the device_class of the device."""
-        return self._device_class
+        return (
+            self.tesla_device.device_class
+            if self.tesla_device.device_class in DEVICE_CLASSES
+            else None
+        )
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        attr = self._attributes.copy()
+        if self.tesla_device.type == "charging rate sensor":
+            attr.update(
+                {
+                    "time_left": self.tesla_device.time_left,
+                    "added_range": self.tesla_device.added_range,
+                    "charge_energy_added": self.tesla_device.charge_energy_added,
+                    "charge_current_request": self.tesla_device.charge_current_request,
+                    "charger_actual_current": self.tesla_device.charger_actual_current,
+                    "charger_voltage": self.tesla_device.charger_voltage,
+                }
+            )
+        return attr
 
     async def async_update(self):
         """Update the state from the sensor."""
-        _LOGGER.debug("Updating sensor: %s", self._name)
+        _LOGGER.debug("Updating sensor: %s", self.name)
```

Comment:
id move this log statement to TeslaDevice but just be generic ie "Updating state for: %s" and drop all these redefinitions that are just there for a singls log statement

Reply:
Yah that make sense and I'll change it. But I also don't think this is actually called anymore in normal operation per my prior comment.